crates/primitives/src/transaction/mod.rs

```diff
@@ -554,9 +665,71 @@ impl Decodable for TransactionKind {
 }
 
 /// Signed transaction.
-#[main_codec(no_arbitrary)]
-#[add_arbitrary_tests(rlp, compact)]
-#[derive(Debug, Clone, PartialEq, Eq, Hash, AsRef, Deref, Default)]
+#[derive_arbitrary(compact)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, AsRef, Deref, Default, Serialize, Deserialize)]
+pub struct TransactionSignedNoHash {
+    /// The transaction signature values
+    pub signature: Signature,
+    /// Raw transaction info
+    #[deref]
+    #[as_ref]
+    pub transaction: Transaction,
+}
+
+impl TransactionSignedNoHash {
+    /// Calculates the transaction hash. If used more than once, it's better to convert it to
+    /// [`TransactionSigned`] first.
+    pub fn hash(&self) -> H256 {
+        let mut buf = Vec::new();
+        self.transaction.encode_with_signature(&self.signature, &mut buf, false);
+        keccak256(&buf)
+    }
+}
+
+impl Compact for TransactionSignedNoHash {
+    fn to_compact<B>(self, buf: &mut B) -> usize
+    where
+        B: bytes::BufMut + AsMut<[u8]>,
+    {
+        let before = buf.as_mut().len();
+
+        // placeholder for bitflags
+        buf.put_u8(0);
+
+        let sig_bit = self.signature.to_compact(buf) as u8;
+        let tx_bit = self.transaction.to_compact(buf) as u8;
+
+        // replace with actual flags
+        buf.as_mut()[before] = sig_bit | (tx_bit << 1);
+
+        buf.as_mut().len() - before
+    }
+
+    fn from_compact(mut buf: &[u8], _: usize) -> (Self, &[u8]) {
+        let prefix = buf.get_u8() as usize;
+
+        let (signature, buf) = Signature::from_compact(buf, prefix & 1);
+        let (transaction, buf) = Transaction::from_compact(buf, prefix >> 1);
+
+        (TransactionSignedNoHash { signature, transaction }, buf)
+    }
+}
+
+impl From<TransactionSignedNoHash> for TransactionSigned {
+    fn from(tx: TransactionSignedNoHash) -> Self {
+        TransactionSigned::from_transaction_and_signature(tx.transaction, tx.signature)
+    }
```

Comment:
I'd also like a function that does this (`with_hash()` perhaps?)
The reason this is more convenient is because we now have a few layers, imagine you need to go from `TransactionSignedNoHash` to `TransactionSignedRecovered` with only `From` impls.

with dedicated functions you can do `no_hash.with_hash().into_erecovered()`